include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,311 @@
+/**
+ * @file nsga2_impl.hpp
+ * @author Sayan Goswami
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more Information.
+ */
+
+
+#ifndef ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+#define ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+
+#include "nsga2.hpp"
+#include <set>
+#include <vector>
+#include <map>
+
+namespace ens {
+
+inline NSGA2::NSGA2(const size_t populationSize,
+                    const size_t maxGenerations,
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const double epsilon) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    epsilon(epsilon)
+{ /* Nothing to do here. */ }
+
+template<typename MultiobjectiveFunctionType,
+         typename MatType,
+         typename... CallbackTypes>
+std::vector<MatType> NSGA2::Optimize(MultiobjectiveFunctionType& objectives,
+                        MatType& iterate,
+                        CallbackTypes&&... callbacks)
+{
+  std::vector<arma::vec> calculatedObjectives;
+  calculatedObjectives.resize(populationSize);
+
+  std::vector<MatType> population;
+  std::vector<std::vector<size_t> > fronts;
+  std::vector<double> crowdingDistance;
+  std::vector<size_t> ranks;
+
+  bool terminate = false;
+
+  for (size_t i = 0; i < populationSize; i++) {
+    population.push_back(arma::randu<MatType>(iterate.n_rows,
+        iterate.n_cols) + iterate);
+  }
+
+  EvaluateObjectives(population, objectives, calculatedObjectives);
```

Comment:
Is there any reason you chose to put this in it's own function? I think we could just put the for loop here with a comment as to it's purpose?

Reply:
I separated the to avoid repetition; the `EvaluateObjectives` function is used multiple (two) times in the Optimize function.